### PR TITLE
Add a test with dependent case

### PR DESCRIPTION
**Pull request by ohad**
I added a test with dependent case splitting. 
If you uncomment any of the holes, you get the associated errors.
I think these are bugs, but maybe I'm expecting too much from the core!

The corresponding idris2 code is the following (which currently doesn't type-check due to [this](https://github.com/idris-lang/Idris2/issues/2201#issuecomment-996217064) bug in the current core):
```idris
import Data.Fin

%default total

public export
data All : (n : Nat) -> (Fin n -> Type) -> Type where
  Nil : All 0 p
  (::) : (prf : p FZ) -> All n (p . FS) -> All (S n) p

public export
cotupleBin : (m, n : Nat) -> (Fin m -> a) -> (Fin n -> a) -> Fin (m + n) -> a
cotupleBin 0 n p q i = q i
cotupleBin (S k) n p q FZ = p FZ
cotupleBin (S k) n p q (FS i) = cotupleBin k n (p . FS) q i

failing "While processing right hand side of ++. Can't solve constraint between: q and cotupleBin 0 m p q."
  public export
  (++) : forall m, n, p, q.
    All n p -> All m q -> All (n + m) (cotupleBin n m p q)
  [] ++ qrfs = qrfs
  (prf :: prfs) ++ qrfs = prf :: prfs ++ qrfs

-- eta-expanding the argument works:
public export
(++) : forall m, n, p, q.
  All n p -> All m q -> All (n + m) (\i => cotupleBin n m p q i)
[] ++ qrfs = qrfs
(prf :: prfs) ++ qrfs = prf :: (prfs ++ qrfs)

```

### CODE REVIEW

**Comment by edwinb** [comment]
Hmmph, I'd written a comment here and it got eaten. Let's try again...
I'm not entirely surprised by this - there are probably places in the raw type checker where we should be substituting the matched term, but aren't. I'm not massively worried, because it's just scaffolding and won't be used in practice, but on the other hand it would be nice and perhaps helpful for debugging later if it was reliable. I would like to have a thing that does complete re-checks of elaborated programs.

I'll mention this at the next meeting and see if anyone is interested in working on the checker for `Raw` terms.

Also, I wonder if this is worth merging in a slightly different form, in a subdirectory for not-yet-working tests?

**Comment by ohad** [comment]
I'm at ICFP, might not make it to today's meeting.
I can take a look at the checker, but might need some help (@mjustus ?)

I didn't create a separate sub-directory, so that tests maintain their name until they're eventually supported.
Is that acceptable?

**Comment by mjustus** [comment]
Here is a reduced test case for `cotupleBin`:

```
data Bool : Type where {
   True : Bool
 | False : Bool
}

-- works

T1 : pi b : Bool. Type;

test1 : pi x : Bool. pi t : T1 x. T1 x;
test1 = lam x. lam t.
  case x of
      True => t
    | False => t
;

-- fails

data T2 : pi b : Bool. Type where {      
}

test2 : pi x : Bool. pi t : T2 x. T2 x;
test2 = lam x. lam t.
  case x of
      True => t
    | False => t
;
```

Removing the conversion check in `replace'`

https://github.com/edwinb/Yaffle/blob/d7f1abc672d8fb31f44885d6537c3fb47fc9b472/src/Core/Evaluate.idr#L54-L62

and always running `repSub tm` makes type checking suceed.

File: tests/tt/unify004/all.tt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+           pi prfs : All n (lam i . p (FS n i)) .
+           All (S n) p
+}
+
+cotupleBin : pi a : Type . pi m : Nat . pi n : Nat .
+  pi p : (pi i : Fin m . a) . pi q : (pi i : Fin n . a) .
+  pi i : Fin (plus m n) . a;
+cotupleBin = lam a . lam m . lam n . lam p . lam q . lam i .
+  case m of
+      Z => ?h0 -- (q i) -- gives: Can't convert case m[4] of [???, ???] with n[3]
```
Comment (gallais) [inline tests/tt/unify004/all.tt:30]: Other PR gives me:
```
Uncaught error: Main:30:15--30:16:Can't convert case m[4] of [Z => n[3], S x' => (S (plus x'[0] n[4]))] with n[3]
```

**Comment by edwinb** [comment]
Curiously, the added test works for me. Is it supposed to? But the minimised example from @mjustus gives the error I expect. Looks like the scrutinee isn't being substituted. For the sake of tidiness I will fix this...

**Comment by gallais** [comment]
> Curiously, the added test works for me. Is it supposed to? 

The bits that don't are commented out and replaced with holes :D

**Comment by edwinb** [comment]
> > Curiously, the added test works for me. Is it supposed to?
> 
> The bits that don't are commented out and replaced with holes :D

Oh yes. I knew that. Just testing, etc.

**Comment by edwinb** [comment]
The problem appears to be that we don't update the types in the environment to take account of the new knowledge. So this isn't really surprising. I'm not sure how to do this efficiently, though maybe that's not so important for this part of the system.

**Pull request merged**